channeldb/graph.go

```diff
@@ -1803,44 +1823,51 @@ func (l *LightningNode) ForEachChannel(tx *bolt.Tx,
 		// as its prefix. This indicates that we've stepped over into
 		// another node's edges, so we can terminate our scan.
 		edgeCursor := edges.Cursor()
-		for nodeEdge, edgeInfo := edgeCursor.Seek(nodeStart[:]); bytes.HasPrefix(nodeEdge, nodePub); nodeEdge, edgeInfo = edgeCursor.Next() {
-			// If the prefix still matches, then the value is the
-			// raw edge information. So we can now serialize the
-			// edge info and fetch the outgoing node in order to
-			// retrieve the full channel edge.
-			edgeReader := bytes.NewReader(edgeInfo)
-			toEdgePolicy, err := deserializeChanEdgePolicy(edgeReader, nodes)
+		for nodeEdge, _ := edgeCursor.Seek(nodeStart[:]); bytes.HasPrefix(nodeEdge, nodePub); nodeEdge, _ = edgeCursor.Next() {
+			// If the prefix still matches, the channel id is
+			// returned in nodeEdge. Channel id is used to lookup
+			// the node at the other end of the channel and both
+			// edge policies.
+			chanID := nodeEdge[33:]
+			edgeInfo, err := fetchChanEdgeInfo(edgeIndex, chanID)
 			if err != nil {
 				return err
 			}
-			toEdgePolicy.db = l.db
-			toEdgePolicy.Node.db = l.db
 
-			chanID := nodeEdge[33:]
-			edgeInfo, err := fetchChanEdgeInfo(edgeIndex, chanID)
-			if err != nil {
+			otherNode := edgeInfo.OtherNodeKeyBytes(nodePub)
+
+			outgoingEdgePolicy, err := fetchChanEdgePolicy(
```

Comment:
Can we use `fetchChanEdgePolicies` to fetch both at the same time?

Reply:
I would say readability wins the race until the extra read can be proven a performance bottleneck.

Reply:
Converted to `fetchChanEdgePolicies`. Some new logic needed to figure out what is node 1 and 2, but no duplication anymore.

Reply:
Hm, more added logic than expected, so we could also keep the existing way of doing it. Up to you.